src/librustc_typeck/check/mod.rs

```diff
@@ -3311,6 +3347,20 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             _ => self.warn_if_unreachable(expr.id, expr.span, "expression")
         }
 
+        // Non-lvalues that produce values of type `!` indicate a
```

Comment:
Yeah, that's interesting. I agree the lvalue rule may be a bit overly broad, although it's not obvious to me that this is wrong. For example, consider `x: &!` -- referencing `*x` does not *cause* divergence, rather it's supposed to just *never happen*. That is, the fact that `x` has type `&!` and you are able to access it reflects the fact that this code is supposed to be dead. The same is true of dereferencing a pointer of type `*const !`, I would argue. It does seem like it would make sense to issue *some* sort of warning here, though!

In any case, failing to set the "diverges!" flag then only affects unreachable code warnings and whether something like `let x: ! = Some(*(ptr: *const !));` is legal (it would not be). Put another way, the diverges flag is a conservative approximation, and it should always be safe to set it *less* (right?).

In any case, I have not yet understood your alternative proposal, so let me try to read and process that.